File: lib/service/storage/shopping_list_storage.dart
import 'dart:async';

import 'package:go_list/model/collections/shopping_list_collection.dart';
import 'package:go_list/model/settings.dart';
import 'package:go_list/model/shopping_list.dart';
import 'package:go_list/service/storage/provider/local_storage_provider.dart';
import 'package:go_list/service/storage/provider/remote_storage_provider.dart';

class ShoppingListStorage {
  final LocalStorageProvider localStorageProvider;
  final RemoteStorageProvider remoteStorageProvider;

  ShoppingListStorage(this.localStorageProvider, this.remoteStorageProvider);

  ShoppingListCollection loadShoppingListsFromLocalStorage() {
    ShoppingListCollection shoppingListsFromLocalStorage =
        localStorageProvider.loadShoppingLists();
    shoppingListsFromLocalStorage.cleanup();
    return shoppingListsFromLocalStorage;
  }

  ShoppingList loadShoppingListFromLocalStorage(String shoppingListId) {
    ShoppingList shoppingListsFromLocalStorage =
        localStorageProvider.loadShoppingList(shoppingListId);
    shoppingListsFromLocalStorage.items.removeItemsDeletedSinceTenDays();
    return shoppingListsFromLocalStorage;
  }

  Stream<ShoppingListCollection> loadShoppingLists() async* {
    ShoppingListCollection shoppingListsFromLocalStorage =
        loadShoppingListsFromLocalStorage();
    yield shoppingListsFromLocalStorage;

    ShoppingListCollection shoppingListsFromRemoteStorage =
        (await remoteStorageProvider.loadShoppingLists());
    shoppingListsFromRemoteStorage.cleanup();

    if (!shoppingListsFromRemoteStorage.equals(shoppingListsFromLocalStorage)) {
      ShoppingListCollection merged =
          shoppingListsFromLocalStorage.merge(shoppingListsFromRemoteStorage);

      // update local storage with changes from remote
      merged.entries.forEach(localStorageProvider.upsertShoppingList);
      // update remote storage with changes from local (e.g. when device was offline during change)
      merged.entries.forEach(remoteStorageProvider.upsertShoppingList);

      yield merged;
    }
  }

  Future<Stream<ShoppingList>> listenForChanges(String shoppingListId) async {
    return (await remoteStorageProvider.listenForChanges(shoppingListId))
        .map((shoppingList) {
      ShoppingList mergedWithLocal =
          loadShoppingListFromLocalStorage(shoppingList.id).merge(shoppingList)
              as ShoppingList;
      localStorageProvider.upsertShoppingList(mergedWithLocal);
      return mergedWithLocal;
    });
  }

  Future<void> upsertShoppingList(ShoppingList shoppingList) async {
    localStorageProvider.upsertShoppingList(shoppingList);
    await remoteStorageProvider.upsertShoppingList(shoppingList);
  }

  void saveSettings(Settings settings) {
    localStorageProvider.saveSettings(settings);
  }

  Settings? loadSettings() {
    return localStorageProvider.loadSettings();
  }

  Future<void> deleteShoppingList(String shoppingListId) async {
    localStorageProvider.deleteShoppingList(shoppingListId);
    await remoteStorageProvider.deleteShoppingList(shoppingListId);
  }
}

File: svelte-golist/src/lib/services/storage/ShoppingListStorage.ts
import { ShoppingList } from '../../models/ShoppingList.js';
import { ShoppingListCollection } from '../../models/collections/ShoppingListCollection.js';
import { Settings } from '../../models/Settings.js';
import { LocalStorageProvider } from './LocalStorageProvider.js';
import { RemoteStorageProvider } from './RemoteStorageProvider.js';

export class ShoppingListStorage {
  private localStorageProvider: LocalStorageProvider;
  private remoteStorageProvider: RemoteStorageProvider;

  constructor(localStorageProvider: LocalStorageProvider, remoteStorageProvider: RemoteStorageProvider) {
    this.localStorageProvider = localStorageProvider;
    this.remoteStorageProvider = remoteStorageProvider;
  }

  loadShoppingListsFromLocalStorage(): ShoppingListCollection {
    const shoppingListsFromLocalStorage = this.localStorageProvider.loadShoppingLists();
    shoppingListsFromLocalStorage.cleanup();
    return shoppingListsFromLocalStorage;
  }

  loadShoppingListFromLocalStorage(shoppingListId: string): ShoppingList {
    const shoppingList = this.localStorageProvider.loadShoppingList(shoppingListId);
    shoppingList.items.removeItemsDeletedSinceDays();
    return shoppingList;
  }

  async* loadShoppingLists(): AsyncGenerator<ShoppingListCollection> {
    const shoppingListsFromLocalStorage = this.loadShoppingListsFromLocalStorage();
    yield shoppingListsFromLocalStorage;

    try {
      const shoppingListsFromRemoteStorage = await this.remoteStorageProvider.loadShoppingLists();
      shoppingListsFromRemoteStorage.cleanup();

      if (!shoppingListsFromRemoteStorage.equals(shoppingListsFromLocalStorage)) {
        const merged = shoppingListsFromLocalStorage.merge(shoppingListsFromRemoteStorage) as ShoppingListCollection;

        // Update local storage with changes from remote
        for (const entry of merged.entries) {
          this.localStorageProvider.upsertShoppingList(entry);
        }
        
        // Update remote storage with changes from local
        for (const entry of merged.entries) {
          await this.remoteStorageProvider.upsertShoppingList(entry);
        }

        yield merged;
      }
    } catch (error) {
      console.error('Failed to sync with remote storage:', error);
      // Continue with local data only
    }
  }

  async listenForChanges(shoppingListId: string): Promise<ReadableStream<ShoppingList>> {
    const remoteStream = await this.remoteStorageProvider.listenForChanges(shoppingListId);
    
    return new ReadableStream({
      start: (controller) => {
        const reader = remoteStream.getReader();
        
        const pump = async () => {
          try {
            while (true) {
              const { done, value } = await reader.read();
              if (done) break;
              
              const mergedWithLocal = this.loadShoppingListFromLocalStorage(value.id)
                .merge(value) as ShoppingList;
              this.localStorageProvider.upsertShoppingList(mergedWithLocal);
              controller.enqueue(mergedWithLocal);
            }
          } catch (error) {
            controller.error(error);
          } finally {
            controller.close();
          }
        };
        
        pump();
      }
    });
  }

  async upsertShoppingList(shoppingList: ShoppingList): Promise<void> {
    this.localStorageProvider.upsertShoppingList(shoppingList);
    await this.remoteStorageProvider.upsertShoppingList(shoppingList);
  }

  saveSettings(settings: Settings): void {
    this.localStorageProvider.saveSettings(settings);
  }

  loadSettings(): Settings | null {
    return this.localStorageProvider.loadSettings();
  }

  async deleteShoppingList(shoppingListId: string): Promise<void> {
    this.localStorageProvider.deleteShoppingList(shoppingListId);
    await this.remoteStorageProvider.deleteShoppingList(shoppingListId);
  }
}